In [1]:
%run outboard_scrape.py

In [48]:
from bs4 import BeautifulSoup
from urllib.request import Request, urlopen
import re
import pandas as pd
import time
from itertools import chain
import random

In [3]:
pd.options.display.max_rows = 999

In [8]:
site= "https://www.metacritic.com/game/switch/divinity-original-sin-ii---definitive-edition/user-reviews?sort-by=most-helpful&num_items=100"
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)
print(soup.prettify)



<bound method Tag.prettify of <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "https://www.w3.org/TR/html4/strict.dtd">
<html xml:lang="en">
<head>
<title>Read User Reviews and Submit your own for Divinity: Original Sin II - Definitive Edition on Switch - Metacritic</title>
<meta content="text/html; charset=utf-8" http-equiv="content-type"/><script type="text/javascript">window.NREUM||(NREUM={}),__nr_require=function(e,n,t){function r(t){if(!n[t]){var o=n[t]={exports:{}};e[t][0].call(o.exports,function(n){var o=e[t][1][n];return r(o||n)},o,o.exports)}return n[t].exports}if("function"==typeof __nr_require)return __nr_require;for(var o=0;o<t.length;o++)r(t[o]);return r}({1:[function(e,n,t){function r(){}function o(e,n,t){return function(){return i(e,[c.now()].concat(u(arguments)),n?null:this,t),n?void 0:this}}var i=e("handle"),a=e(3),u=e(4),f=e("ee").get("tracer"),c=e("loader"),s=NREUM;"undefined"==typeof window.newrelic&&(newrelic=s);var p=["setPageViewName","setCustomAttribute","setE

In [9]:
title = str(soup.find('h1'))
title = title[4:-5]
title

'Divinity: Original Sin II - Definitive Edition'

In [8]:
Game_Title.append(title)

## DF for Surprise

In [306]:
## Testing the scraping

site= "https://www.metacritic.com/game/switch/divinity-original-sin-ii---definitive-edition/user-reviews?sort-by=most-helpful&num_items=100"
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

UserID = []
Game_Title = []
BracketRating = []

userdivs = soup.findAll('div', class_='name')
ratingdivs = soup.findAll('div', class_='user')
ratingdivs=ratingdivs[2:]

for user in userdivs:
    user = str(user.find('a'))
    user = user[15:]
    user = re.findall("\w+", user)
    user = user[0]
    UserID.append(user)
    
for user in UserID:
    Game_Title.append(title)
    
for rating in ratingdivs:
    rating = re.findall("\d\d?",str(rating))
    rating = list(map(int, rating))
    BracketRating.append(rating)
    
Rating = []
for sublist in BracketRating:
    for item in sublist:
        Rating.append(item)
        
Userdf = pd.DataFrame([UserID, Game_Title, Rating]).transpose()
Userdf.columns = ['UserID', 'Game_Title', 'Rating']
Userdf

UserID                                      Game_Title Rating
0       ChadTyler  Divinity: Original Sin II - Definitive Edition     10
1           Dazan  Divinity: Original Sin II - Definitive Edition     10
2          A_coat  Divinity: Original Sin II - Definitive Edition      8
3      pgualtieri  Divinity: Original Sin II - Definitive Edition     10
4       Artezzany  Divinity: Original Sin II - Definitive Edition     10
5      kazuki1866  Divinity: Original Sin II - Definitive Edition     10
6          Aaargh  Divinity: Original Sin II - Definitive Edition     10
7         Arigoki  Divinity: Original Sin II - Definitive Edition     10
8         dpruttz  Divinity: Original Sin II - Definitive Edition     10
9   SlimeKnight33  Divinity: Original Sin II - Definitive Edition     10
10   evildoomfoot  Divinity: Original Sin II - Definitive Edition     10
11          Rpg89  Divinity: Original Sin II - Definitive Edition     10
12   live2winalot  Divinity: Original Sin II - Definitive Edition     10
13  Bellonintendo  Divinity: Original Sin II - Definitive Edition     10
14    Swipebungee  Divinity: Original Sin II - Definitive Edition     10
15       OlorynEx  Divinity: Original Sin II - Definitive Edition      9
16        Bandaid  Divinity: Original Sin II - Definitive Edition     10
17      houserp85  Divinity: Original Sin II - Definitive Edition     10
18         Kabaal  Divinity: Original Sin II - Definitive Edition     10
19      Perkstyle  Divinity: Original Sin II - Definitive Edition     10
20        Daebiya  Divinity: Original Sin II - Definitive Edition     10
21          Bnois  Divinity: Original Sin II - Definitive Edition      4
22      TruckHeck  Divinity: Original Sin II - Definitive Edition      2
23         Hikage  Divinity: Original Sin II - Definitive Edition      4
24    Thwompthing  Divinity: Original Sin II - Definitive Edition      0
25     RagnarX213  Divinity: Original Sin II - Definitive Edition      9
26       Mandiven  Divinity: Original Sin II - Definitive Edition      9
27            Ama  Divinity: Original Sin II - Definitive Edition     10
28     YTG_GAMING  Divinity: Original Sin II - Definitive Edition     10
29    MrGoodbarHD  Divinity: Original Sin II - Definitive Edition     10

In [236]:
len(UserID)

30

In [237]:
len(Game_Title)

30

In [238]:
len(Rating)

30

### Scraping the URLs & Game Titles!

In [18]:
## Scraping the first page

site= "https://www.metacritic.com/browse/games/release-date/available/switch/metascore"
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

BracketURLs = []
urldivs = soup.findAll('div', class_='basic_stat')
for url in urldivs:
    url = str(url.find('a'))
    url = url[9:]
    url = re.findall('.+?(?=">\n)', url)
    BracketURLs.append(url)

BracketURLs = BracketURLs[0::2]
BracketURLs = BracketURLs[:len(BracketURLs)-2]

URL = []
for sublist in BracketURLs:
    for item in sublist:
        URL.append(item)
        
BracketTitle = []
titledivs = soup.findAll('div', class_='basic_stat product_title')
titledivs = titledivs[0:200]

for title in titledivs:
    title = str(title.find('a'))
    title = re.findall('(?<=\\n)(.*?)(?=\\n)', title)
    title = [title.lstrip() for title in title]
    BracketTitle.append(title)
    
Title = []
for sublist in BracketTitle:
    for item in sublist:
        Title.append(item)

In [12]:
URL

['/game/switch/the-legend-of-zelda-breath-of-the-wild',
 '/game/switch/super-mario-odyssey',
 '/game/switch/divinity-original-sin-ii---definitive-edition',
 '/game/switch/undertale',
 '/game/switch/super-smash-bros-ultimate',
 '/game/switch/celeste',
 '/game/switch/bayonetta-2',
 '/game/switch/mario-kart-8-deluxe',
 '/game/switch/inside',
 '/game/switch/retro-city-rampage-dx',
 '/game/switch/sonic-mania-plus',
 '/game/switch/steamworld-heist-ultimate-edition',
 '/game/switch/dragon-quest-xi-s-echoes-of-an-elusive-age---definitive-edition',
 '/game/switch/shovel-knight-treasure-trove',
 '/game/switch/bastion',
 '/game/switch/hollow-knight',
 '/game/switch/bayonetta-+-bayonetta-2',
 '/game/switch/ori-and-the-blind-forest-definitive-edition',
 '/game/switch/dead-cells',
 '/game/switch/into-the-breach',
 '/game/switch/fire-emblem-three-houses',
 '/game/switch/okami-hd',
 '/game/switch/mark-of-the-ninja-remastered',
 '/game/switch/towerfall',
 '/game/switch/nuclear-throne',
 '/game/switch/f

In [13]:
Title

['The Legend of Zelda: Breath of the Wild',
 'Super Mario Odyssey',
 'Divinity: Original Sin II - Definitive Edition',
 'Undertale',
 'Super Smash Bros. Ultimate',
 'Celeste',
 'Bayonetta 2',
 'Mario Kart 8 Deluxe',
 'INSIDE',
 'Retro City Rampage DX',
 'Sonic Mania Plus',
 'SteamWorld Heist: Ultimate Edition',
 'Dragon Quest XI S: Echoes of an Elusive Age - Definitive Edition',
 'Shovel Knight: Treasure Trove',
 'Bastion',
 'Hollow Knight',
 'Bayonetta + Bayonetta 2',
 'Ori and the Blind Forest: Definitive Edition',
 'Dead Cells',
 'Into the Breach',
 'Fire Emblem: Three Houses',
 'Okami HD',
 'Mark of the Ninja: Remastered',
 'TowerFall',
 'Nuclear Throne',
 'FAR: Lone Sails',
 'Dream Daddy: A Dad Dating Simulator',
 'Hyper Light Drifter',
 'What Remains of Edith Finch',
 'Gone Home',
 'Super Mario Maker 2',
 'Downwell',
 'Diablo III: Eternal Collection',
 'SteamWorld Dig 2',
 'Ikaruga',
 'Cave Story +',
 'Deemo',
 'Dragon Ball FighterZ',
 'Stardew Valley',
 'Guacamelee! 2',
 'Enter 

In [19]:
URLdf = pd.DataFrame([Title, URL]).transpose()
URLdf.columns = ['Title', 'URL']
URLdf

Title  \
0              The Legend of Zelda: Breath of the Wild   
1                                  Super Mario Odyssey   
2       Divinity: Original Sin II - Definitive Edition   
3                                            Undertale   
4                           Super Smash Bros. Ultimate   
5                                              Celeste   
6                                          Bayonetta 2   
7                                  Mario Kart 8 Deluxe   
8                                               INSIDE   
9                                Retro City Rampage DX   
10                                    Sonic Mania Plus   
11   Dragon Quest XI S: Echoes of an Elusive Age - ...   
12                  SteamWorld Heist: Ultimate Edition   
13                       Shovel Knight: Treasure Trove   
14                                             Bastion   
15                                       Hollow Knight   
16                             Bayonetta + Bayonetta 2   
17        Ori and the Blind Forest: Definitive Edition   
18                                          Dead Cells   
19                                     Into the Breach   
20                           Fire Emblem: Three Houses   
21                                            Okami HD   
22                       Mark of the Ninja: Remastered   
23                                           TowerFall   
24                                      Nuclear Throne   
25                                     FAR: Lone Sails   
26                 Dream Daddy: A Dad Dating Simulator   
27                                 Hyper Light Drifter   
28                         What Remains of Edith Finch   
29                                           Gone Home   
30                                 Super Mario Maker 2   
31                                            Downwell   
32                      Diablo III: Eternal Collection   
33                                    SteamWorld Dig 2   
34                                             Ikaruga   
35                                        Cave Story +   
36                                               Deemo   
37                                Dragon Ball FighterZ   
38                                      Stardew Valley   
39                                       Guacamelee! 2   
40                                   Enter the Gungeon   
41                                             Cuphead   
42               The Legend of Zelda: Link's Awakening   
43                                        Astral Chain   
44                             Trine 2: Complete Story   
45                                             Severed   
46                                         Axiom Verge   
47                              Katamari Damacy REROLL   
48                                              Owlboy   
49             Super Mega Baseball 2: Ultimate Edition   
50                                         Velocity 2X   
51                                         Iconoclasts   
52                                       Rocket League   
53   Crypt of the NecroDancer: Nintendo Switch Edition   
54                                       The Messenger   
55                Donkey Kong Country: Tropical Freeze   
56                                   Star Wars Pinball   
57                  Monster Boy and the Cursed Kingdom   
58                                          Transistor   
59                           Minecraft: Switch Edition   
60                      Shantae and the Pirate's Curse   
61        Guacamelee! Super Turbo Championship Edition   
62                             Assault Android Cactus+   
63                         Sid Meier's Civilization VI   
64                             Return of the Obra Dinn   
65                                            Warframe   
66                                        Tangle Tower   
67                               AI: The Somnium Files   
68         BlazBlue: Central Fiction - Special 

In [901]:
URLdf['Title']

0                 The Legend of Zelda: Breath of the Wild
1                                     Super Mario Odyssey
2          Divinity: Original Sin II - Definitive Edition
3                                               Undertale
4                              Super Smash Bros. Ultimate
5                                                 Celeste
6                                             Bayonetta 2
7                                     Mario Kart 8 Deluxe
8                                                  INSIDE
9                                   Retro City Rampage DX
10                                       Sonic Mania Plus
11                     SteamWorld Heist: Ultimate Edition
12      Dragon Quest XI S: Echoes of an Elusive Age - ...
13                          Shovel Knight: Treasure Trove
14                                                Bastion
15                                          Hollow Knight
16                                Bayonetta + Bayonetta 2
17           O

In [20]:
## Scraping the other pages

site = 'https://www.metacritic.com/browse/games/release-date/available/switch/metascore?page=1'
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

BracketURLs = []
urldivs = soup.findAll('div', class_='basic_stat')
for url in urldivs:
    url = str(url.find('a'))
    url = url[9:]
    url = re.findall('.+?(?=">\n)', url)
    BracketURLs.append(url)

BracketURLs = BracketURLs[0::2]
BracketURLs = BracketURLs[:len(BracketURLs)-2]

for sublist in BracketURLs:
    for item in sublist:
        URL.append(item)
        
BracketTitle = []
titledivs = soup.findAll('div', class_='basic_stat product_title')
titledivs = titledivs[0:200]

for title in titledivs:
    title = str(title.find('a'))
    title = re.findall('(?<=\\n)(.*?)(?=\\n)', title)
    title = [title.lstrip() for title in title]
    BracketTitle.append(title)
    
for sublist in BracketTitle:
    for item in sublist:
        Title.append(item)

In [532]:
len(URL)

400

In [21]:
site = 'https://www.metacritic.com/browse/games/release-date/available/switch/metascore?page=2'
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

BracketURLs = []
urldivs = soup.findAll('div', class_='basic_stat')
for url in urldivs:
    url = str(url.find('a'))
    url = url[9:]
    url = re.findall('.+?(?=">\n)', url)
    BracketURLs.append(url)

BracketURLs = BracketURLs[0::2]
BracketURLs = BracketURLs[:len(BracketURLs)-2]

for sublist in BracketURLs:
    for item in sublist:
        URL.append(item)
        
BracketTitle = []
titledivs = soup.findAll('div', class_='basic_stat product_title')
titledivs = titledivs[0:200]

for title in titledivs:
    title = str(title.find('a'))
    title = re.findall('(?<=\\n)(.*?)(?=\\n)', title)
    title = [title.lstrip() for title in title]
    BracketTitle.append(title)
    
for sublist in BracketTitle:
    for item in sublist:
        Title.append(item)

In [22]:
site = 'https://www.metacritic.com/browse/games/release-date/available/switch/metascore?page=3'
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

BracketURLs = []
urldivs = soup.findAll('div', class_='basic_stat')
for url in urldivs:
    url = str(url.find('a'))
    url = url[9:]
    url = re.findall('.+?(?=">\n)', url)
    BracketURLs.append(url)

BracketURLs = BracketURLs[0::2]
BracketURLs = BracketURLs[:len(BracketURLs)-2]

for sublist in BracketURLs:
    for item in sublist:
        URL.append(item)
        
BracketTitle = []
titledivs = soup.findAll('div', class_='basic_stat product_title')
titledivs = titledivs[0:200]

for title in titledivs:
    title = str(title.find('a'))
    title = re.findall('(?<=\\n)(.*?)(?=\\n)', title)
    title = [title.lstrip() for title in title]
    BracketTitle.append(title)
    
for sublist in BracketTitle:
    for item in sublist:
        Title.append(item)

In [23]:
site = 'https://www.metacritic.com/browse/games/release-date/available/switch/metascore?page=4'
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

BracketURLs = []
urldivs = soup.findAll('div', class_='basic_stat')
for url in urldivs:
    url = str(url.find('a'))
    url = url[9:]
    url = re.findall('.+?(?=">\n)', url)
    BracketURLs.append(url)

BracketURLs = BracketURLs[0::2]
BracketURLs = BracketURLs[:len(BracketURLs)-2]

for sublist in BracketURLs:
    for item in sublist:
        URL.append(item)
        
BracketTitle = []
titledivs = soup.findAll('div', class_='basic_stat product_title')
titledivs = titledivs[0:200]

for title in titledivs:
    title = str(title.find('a'))
    title = re.findall('(?<=\\n)(.*?)(?=\\n)', title)
    title = [title.lstrip() for title in title]
    BracketTitle.append(title)
    
for sublist in BracketTitle:
    for item in sublist:
        Title.append(item)

In [24]:
site = 'https://www.metacritic.com/browse/games/release-date/available/switch/metascore?page=5'
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

BracketURLs = []
urldivs = soup.findAll('div', class_='basic_stat')
for url in urldivs:
    url = str(url.find('a'))
    url = url[9:]
    url = re.findall('.+?(?=">\n)', url)
    BracketURLs.append(url)

BracketURLs = BracketURLs[0::2]
BracketURLs = BracketURLs[:len(BracketURLs)-2]

for sublist in BracketURLs:
    for item in sublist:
        URL.append(item)
        
BracketTitle = []
titledivs = soup.findAll('div', class_='basic_stat product_title')
titledivs = titledivs[0:200]

for title in titledivs:
    title = str(title.find('a'))
    title = re.findall('(?<=\\n)(.*?)(?=\\n)', title)
    title = [title.lstrip() for title in title]
    BracketTitle.append(title)
    
for sublist in BracketTitle:
    for item in sublist:
        Title.append(item)

In [26]:
site = 'https://www.metacritic.com/browse/games/release-date/available/switch/metascore?page=6'
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

BracketURLs = []
urldivs = soup.findAll('div', class_='basic_stat')
for url in urldivs:
    url = str(url.find('a'))
    url = url[9:]
    url = re.findall('.+?(?=">\n)', url)
    BracketURLs.append(url)

BracketURLs = BracketURLs[0::2]
BracketURLs = BracketURLs[:len(BracketURLs)-2]

for sublist in BracketURLs:
    for item in sublist:
        URL.append(item)
        
BracketTitle = []
titledivs = soup.findAll('div', class_='basic_stat product_title')
titledivs = titledivs[0:200]

for title in titledivs:
    title = str(title.find('a'))
    title = re.findall('(?<=\\n)(.*?)(?=\\n)', title)
    title = [title.lstrip() for title in title]
    BracketTitle.append(title)
    
for sublist in BracketTitle:
    for item in sublist:
        Title.append(item)

In [27]:
site = 'https://www.metacritic.com/browse/games/release-date/available/switch/metascore?page=7'
hdr = {'User-Agent': 'Mozilla/5.0'}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

BracketURLs = []
urldivs = soup.findAll('div', class_='basic_stat')
for url in urldivs:
    url = str(url.find('a'))
    url = url[9:]
    url = re.findall('.+?(?=">\n)', url)
    BracketURLs.append(url)

BracketURLs = BracketURLs[0::2]
BracketURLs = BracketURLs[:len(BracketURLs)-2]

for sublist in BracketURLs:
    for item in sublist:
        URL.append(item)
        
BracketTitle = []
titledivs = soup.findAll('div', class_='basic_stat product_title')
titledivs = titledivs[0:200]

for title in titledivs:
    title = str(title.find('a'))
    title = re.findall('(?<=\\n)(.*?)(?=\\n)', title)
    title = [title.lstrip() for title in title]
    BracketTitle.append(title)
    
for sublist in BracketTitle:
    for item in sublist:
        Title.append(item)

In [28]:
URLdf = pd.DataFrame([Title, URL]).transpose()
URLdf.columns = ['Title', 'URL']
URLdf

Title  \
0               The Legend of Zelda: Breath of the Wild   
1                                   Super Mario Odyssey   
2        Divinity: Original Sin II - Definitive Edition   
3                                             Undertale   
4                            Super Smash Bros. Ultimate   
5                                               Celeste   
6                                           Bayonetta 2   
7                                   Mario Kart 8 Deluxe   
8                                                INSIDE   
9                                 Retro City Rampage DX   
10                                     Sonic Mania Plus   
11    Dragon Quest XI S: Echoes of an Elusive Age - ...   
12                   SteamWorld Heist: Ultimate Edition   
13                        Shovel Knight: Treasure Trove   
14                                              Bastion   
15                                        Hollow Knight   
16                              Bayonetta + Bayonetta 2   
17         Ori and the Blind Forest: Definitive Edition   
18                                           Dead Cells   
19                                      Into the Breach   
20                            Fire Emblem: Three Houses   
21                                             Okami HD   
22                        Mark of the Ninja: Remastered   
23                                            TowerFall   
24                                       Nuclear Throne   
25                                      FAR: Lone Sails   
26                  Dream Daddy: A Dad Dating Simulator   
27                                  Hyper Light Drifter   
28                          What Remains of Edith Finch   
29                                            Gone Home   
30                                  Super Mario Maker 2   
31                                             Downwell   
32                       Diablo III: Eternal Collection   
33                                     SteamWorld Dig 2   
34                                              Ikaruga   
35                                         Cave Story +   
36                                                Deemo   
37                                 Dragon Ball FighterZ   
38                                       Stardew Valley   
39                                        Guacamelee! 2   
40                                    Enter the Gungeon   
41                                              Cuphead   
42                The Legend of Zelda: Link's Awakening   
43                                         Astral Chain   
44                              Trine 2: Complete Story   
45                                              Severed   
46                                          Axiom Verge   
47                               Katamari Damacy REROLL   
48                                               Owlboy   
49              Super Mega Baseball 2: Ultimate Edition   
50                                          Velocity 2X   
51                                          Iconoclasts   
52                                        Rocket League   
53    Crypt of the NecroDancer: Nintendo Switch Edition   
54                                        The Messenger   
55                 Donkey Kong Country: Tropical Freeze   
56                                    Star Wars Pinball   
57                   Monster Boy and the Cursed Kingdom   
58                                           Transistor   
59                            Minecraft: Switch Edition   
60                       Shantae and the Pirate's Curse   
61         Guacamelee! Super Turbo Championship Edition   
62                              Assault Android Cactus+   
63                          Sid Meier's Civilization VI   
64                              Return of the Obra Dinn   
65                                             Warframe   
66                                         Tangle Tower   
67                                AI: 

In [679]:
URLdf.to_csv('AllGames.csv')

In [49]:
def rand_user_agent():
    user_agent_list = [
       #Chrome
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36',
        'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36',
        'Mozilla/5.0 (Windows NT 5.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36',
        'Mozilla/5.0 (Windows NT 6.2; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36',
        'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.157 Safari/537.36',
        'Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36',
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36',
        'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36',
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.87 Safari/537.36',
        'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.87 Safari/537.36',
        #Firefox
        'Mozilla/4.0 (compatible; MSIE 9.0; Windows NT 6.1)',
        'Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; rv:11.0) like Gecko',
        'Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; WOW64; Trident/5.0)',
        'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko',
        'Mozilla/5.0 (Windows NT 6.2; WOW64; Trident/7.0; rv:11.0) like Gecko',
        'Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; rv:11.0) like Gecko',
        'Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.0; Trident/5.0)',
        'Mozilla/5.0 (Windows NT 6.3; WOW64; Trident/7.0; rv:11.0) like Gecko',
        'Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)',
        'Mozilla/5.0 (Windows NT 6.1; Win64; x64; Trident/7.0; rv:11.0) like Gecko',
        'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.1; WOW64; Trident/6.0)',
        'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.1; Trident/6.0)',
        'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729)'
    ]

    return random.choice(user_agent_list)

In [5]:
UserID = []
Game_Title = []
Rating = []

In [8]:
Userdf = pd.DataFrame.from_csv('UserReviews.csv')

/Users/flatironschool/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls
  """Entry point for launching an IPython kernel.


In [9]:
Userdf

UserID                                         Game_Title  \
0            johnardo            The Legend of Zelda: Breath of the Wild   
1            triskyhd            The Legend of Zelda: Breath of the Wild   
2       silver_silver            The Legend of Zelda: Breath of the Wild   
3      EpicLadySponge            The Legend of Zelda: Breath of the Wild   
4              ocidax            The Legend of Zelda: Breath of the Wild   
5              xDeniz            The Legend of Zelda: Breath of the Wild   
6          gintoki125            The Legend of Zelda: Breath of the Wild   
7          Geraldst20            The Legend of Zelda: Breath of the Wild   
8          Sadikusama            The Legend of Zelda: Breath of the Wild   
9              jjakub            The Legend of Zelda: Breath of the Wild   
10      VideoGameDude            The Legend of Zelda: Breath of the Wild   
11          fire_fist            The Legend of Zelda: Breath of the Wild   
12           jardeson            The Legend of Zelda: Breath of the Wild   
13          grungyssj            The Legend of Zelda: Breath of the Wild   
14    AlexRamosQueiro            The Legend of Zelda: Breath of the Wild   
15       Smashmatt202            The Legend of Zelda: Breath of the Wild   
16          Prodromus            The Legend of Zelda: Breath of the Wild   
17         shawndumas            The Legend of Zelda: Breath of the Wild   
18    Fred_GameOnBatt            The Legend of Zelda: Breath of the Wild   
19     PajaritoValdes            The Legend of Zelda: Breath of the Wild   
20           Koppai15            The Legend of Zelda: Breath of the Wild   
21          Kiryu1120            The Legend of Zelda: Breath of the Wild   
22            zorakex            The Legend of Zelda: Breath of the Wild   
23          NInteNoob            The Legend of Zelda: Breath of the Wild   
24       leonardo2360            The Legend of Zelda: Breath of the Wild   
25             Xalusc            The Legend of Zelda: Breath of the Wild   
26        lesley_dias            The Legend of Zelda: Breath of the Wild   
27             altalo            The Legend of Zelda: Breath of the Wild   
28             Volian            The Legend of Zelda: Breath of the Wild   
29           LaurensL            The Legend of Zelda: Breath of the Wild   
30         flujackson            The Legend of Zelda: Breath of the Wild   
31        iusshpandeh            The Legend of Zelda: Breath of the Wild   
32      Schoolboy_hew            The Legend of Zelda: Breath of the Wild   
33           chandori            The Legend of Zelda: Breath of the Wild   
34         joseloko89            The Legend of Zelda: Breath of the Wild   
35         ChromXsome            The Legend of Zelda: Breath of the Wild   
36         The_Legend            The Legend of Zelda: Breath of the Wild   
37       beetleking22            The Legend of Zelda: Breath of the Wild   
38         Nessuno_Im            The Legend of Zelda: Breath of the Wild   
39               pg_x            The Legend of Zelda: Breath of the Wild   
40          someone98            The Legend of Zelda: Breath of the Wild   
41     FireFistYamaan            The Legend of Zelda: Breath of the Wild   
42       JuanLuisGG14            The Legend of Zelda: Breath of the Wild   
43         bimbarabim            The Legend of Zelda: Breath of the Wild   
44                  K            The Legend of Zelda: Breath of the Wild   
45             Fhranz            The Legend of Zelda: Breath of the Wild   
46            yasserb            The Legend of Zelda: Breath of the Wild   
47      darthshadow25            The Legend of Zelda: Breath of the Wild   
48    DefectiveReject            The Legend of Zelda: Breath of the Wild   
49           gaydingo            The Legend of Zelda: Breath of the Wild   
50     StefanGoedhart            The Legend of Zelda: Breath of the Wild   
51      GustavoAndres            The Legend of Zelda: Breath of

In [10]:
UserID = list(Userdf['UserID'])

In [11]:
Game_Title = list(Userdf['Game_Title'])

In [12]:
Rating = list(Userdf['Rating'])

In [13]:
len(UserID)

8445

In [14]:
len(Game_Title)

8445

In [15]:
len(Rating)

8445

In [2154]:
Game_Title[-1]

'Warframe'

In [57]:
scrapeurls = 'https://www.metacritic.com' + URLdf['URL'] + '/user-reviews'

In [38]:
len(scrapeurls)

1258

In [39]:
rating = []

In [672]:
scrapeurls[162]

'https://www.metacritic.com/game/switch/valkyria-chronicles-4/user-reviews'

In [673]:
site= scrapeurls[162] 
# + '?page=19'

user_agent = rand_user_agent()
hdr = {'User-Agent': user_agent}
req = Request(site,headers=hdr)
page = urlopen(req)
soup = BeautifulSoup(page)

userdivs = soup.findAll('div', class_='name')
ratingdivs = soup.findAll('div', class_='user')
ratingdivs=ratingdivs[2:]


for user in userdivs:
    user = str(user.find('a'))[15:]
    user = re.findall('.*?(?=\")',user)[0]
    UserID.append(user)
    
for user in userdivs:
    title = str(soup.find('h1'))
    title = title[4:-5]
    Game_Title.append(title)
    
for rating in ratingdivs:
    rating = re.findall("\d\d?",str(ratingdivs))
    rating = list(map(int, rating)) 

    
Rating.extend(rating)
rating = []

In [674]:
len(Game_Title)

13911

In [675]:
len(Rating)

13911

In [671]:
len(UserID)

13891

In [461]:
Game_Title[-1]

'Dark Souls Remastered'

In [314]:
len(set(Game_Title))

114

In [374]:
len(set(Game_Title))

121

In [676]:
Rating[-11:-1]

[6, 9, 8, 10, 8, 10, 9, 10, 8, 10]

In [677]:
Userdf = pd.DataFrame([UserID, Game_Title, Rating]).transpose()
Userdf.columns = ['UserID', 'Game_Title', 'Rating']
len(Userdf)

13911

In [43]:
scrapeurls

0       metacritic.com/game/switch/the-legend-of-zelda...
1          metacritic.com/game/switch/super-mario-odyssey
2       metacritic.com/game/switch/divinity-original-s...
3                    metacritic.com/game/switch/undertale
4       metacritic.com/game/switch/super-smash-bros-ul...
5                      metacritic.com/game/switch/celeste
6                  metacritic.com/game/switch/bayonetta-2
7          metacritic.com/game/switch/mario-kart-8-deluxe
8                       metacritic.com/game/switch/inside
9        metacritic.com/game/switch/retro-city-rampage-dx
10            metacritic.com/game/switch/sonic-mania-plus
11      metacritic.com/game/switch/dragon-quest-xi-s-e...
12      metacritic.com/game/switch/steamworld-heist-ul...
13      metacritic.com/game/switch/shovel-knight-treas...
14                     metacritic.com/game/switch/bastion
15               metacritic.com/game/switch/hollow-knight
16      metacritic.com/game/switch/bayonetta-+-bayonet...
17      metacr

In [42]:
len(set(Userdf['Game_Title']))

67

In [41]:
set(Userdf['Game_Title'])

{'AI: The Somnium Files',
 'Assault Android Cactus+',
 'Astral Chain',
 'Axiom Verge',
 'Bastion',
 'Bayonetta + Bayonetta 2',
 'Bayonetta 2',
 'Cave Story +',
 'Celeste',
 'Crypt of the NecroDancer: Nintendo Switch Edition',
 'Cuphead',
 'Dead Cells',
 'Deemo',
 'Diablo III: Eternal Collection',
 'Digimon Story Cyber Sleuth: Complete Edition',
 'Divinity: Original Sin II - Definitive Edition',
 'Donkey Kong Country: Tropical Freeze',
 'Downwell',
 'Dragon Ball FighterZ',
 'Dragon Quest XI S: Echoes of an Elusive Age - Definitive Edition',
 'Dream Daddy: A Dad Dating Simulator',
 'Enter the Gungeon',
 'FAR: Lone Sails',
 'Fire Emblem: Three Houses',
 'Gone Home',
 'Guacamelee! 2',
 'Guacamelee! Super Turbo Championship Edition',
 'Hollow Knight',
 'Hyper Light Drifter',
 'INSIDE',
 'Iconoclasts',
 'Ikaruga',
 'Into the Breach',
 'Katamari Damacy REROLL',
 'Mario Kart 8 Deluxe',
 'Mark of the Ninja: Remastered',
 'Minecraft: Switch Edition',
 'Monster Boy and the Cursed Kingdom',
 'Nucl

In [980]:
Userdf = Userdf.dropna()
len(Userdf)

56

In [678]:
Userdf.to_csv('UserReviews.csv')

In [1053]:
import numpy as np
import pandas as pd

from surprise import Dataset, Reader
from surprise import SVD
from surprise import accuracy
from surprise.model_selection import cross_validate, train_test_split

In [1063]:
data = Userdf

## Distribution of Ratings

In [2052]:
Userdf.Rating.value_counts()

10    6290
9      896
8      363
7      333
0      167
6      153
5      143
4       80
1       58
3       43
2       34
Name: Rating, dtype: int64

## Number of reviews per game

In [1066]:
print(Userdf.Game_Title.value_counts().head())
print(Userdf.Game_Title.value_counts().tail())

The Legend of Zelda: Breath of the Wild    3434
Super Mario Odyssey                        1166
Super Smash Bros. Ultimate                  743
Fire Emblem: Three Houses                   604
Mario Kart 8 Deluxe                         316
Name: Game_Title, dtype: int64
Dream Daddy: A Dad Dating Simulator    2
Trine 2: Complete Story                2
Star Wars Pinball                      2
Mark of the Ninja: Remastered          1
Velocity 2X                            1
Name: Game_Title, dtype: int64


## Number of reviews per user

In [1068]:
print(Userdf.UserID.value_counts().head())
print(Userdf.UserID.value_counts().tail())

AllySky        16
bulger_paul    16
nusco          15
Sijun          13
jeffbizloc     12
Name: UserID, dtype: int64
enmanuel1263       1
ImSirEclipse       1
StudentOfTheGam    1
SaydaNeen          1
BlunderBuck        1
Name: UserID, dtype: int64


In [1069]:
# Instantiate a 'Reader' to read in the data so Surprise can use it
reader = Reader(rating_scale=(1, 10))

# The columns must correspond to user id, item id and ratings (in that order).
data = Dataset.load_from_df(Userdf[['UserID', 'Game_Title', 'Rating']], reader)

In [1070]:
trainset, testset = train_test_split(data, test_size=.2)

In [1071]:
final = SVD(n_epochs=20, n_factors=1, biased=True, 
              lr_all=0.005, reg_all=0.06)

In [1072]:
final.fit(trainset)

In [1073]:
predictions = final.test(testset)

In [1074]:
accuracy.rmse(predictions)
accuracy.mae(predictions)

RMSE: 1.7941
MAE:  1.1042


1.1042308295992085

On average, the predictions are 1.79 points off from the actual rating

In [1076]:
final.predict(Userdf['UserID'][55], Userdf['Game_Title'][13])

Prediction(uid='chiscolo', iid='The Legend of Zelda: Breath of the Wild', r_ui=None, est=9.229647325257167, details={'was_impossible': False})

In [1085]:
from collections import defaultdict

from surprise import SVD
from surprise import Dataset


def get_top_n(predictions, n=5):
    '''Return the top-N recommendation for each user from a set of predictions.

    Args:
        predictions(list of Prediction objects): The list of predictions, as
            returned by the test method of an algorithm.
        n(int): The number of recommendation to output for each user. Default
            is 10.

    Returns:
    A dict where keys are user (raw) ids and values are lists of tuples:
        [(raw item id, rating estimation), ...] of size n.
    '''

    # First map the predictions to each user.
    top_n = defaultdict(list)
    for uid, iid, true_r, est, _ in predictions:
        top_n[uid].append((iid, est))

    # Then sort the predictions for each user and retrieve the k highest ones.
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = user_ratings[:n]

    return top_n


# First train an SVD algorithm on the dataset.
trainset = data.build_full_trainset()
algo = SVD()
algo.fit(trainset)

# Than predict ratings for all pairs (u, i) that are NOT in the training set.
testset = trainset.build_anti_testset()
predictions = algo.test(testset)

top_n = get_top_n(predictions, n=5)

# Print the recommended items for each user
for uid, user_ratings in top_n.items():
    print(uid, [iid for (iid, _) in user_ratings])

johnardo ['Enter the Gungeon', 'Super Smash Bros. Ultimate', 'Cuphead', 'Celeste', 'Into the Breach']
triskyhd ['Cuphead', 'Cave Story +', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Stardew Valley']
silver_silver ['Enter the Gungeon', 'Super Mario Odyssey', "The Legend of Zelda: Link's Awakening", 'Super Smash Bros. Ultimate', 'Cave Story +']
EpicLadySponge ['Stardew Valley', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Cave Story +', 'Celeste']
ocidax ['Enter the Gungeon', 'Celeste', "The Legend of Zelda: Link's Awakening", 'Cave Story +', 'Cuphead']
xDeniz ['Enter the Gungeon', 'Dead Cells', 'Cuphead', 'Super Smash Bros. Ultimate', 'Undertale']
gintoki125 ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Diablo III: Eternal Collection', 'Cuphead']
Geraldst20 ['Super Smash Bros. Ultimate', "The Legend of Zelda: Link's Awakening", 'Super Mario Odyssey', 'Enter the Gungeon', 'Mario Kart 8 Deluxe']
Sadikusama ['Super Smash Br

jonnyd_94 ['Super Mario Odyssey', 'Enter the Gungeon', 'Cuphead', "The Legend of Zelda: Link's Awakening", 'What Remains of Edith Finch']
ToshiroTakashi ['Enter the Gungeon', 'Ikaruga', 'Celeste', 'Cuphead', 'Fire Emblem: Three Houses']
durabull ['Enter the Gungeon', 'Cave Story +', 'Stardew Valley', 'Cuphead', 'Celeste']
phenom_cs ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Super Smash Bros. Ultimate', 'Ikaruga', 'Celeste']
maykee79 ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Cuphead', 'Celeste', 'Ikaruga']
Lehtis ['Enter the Gungeon', 'Cuphead', 'What Remains of Edith Finch', 'Cave Story +', 'Super Smash Bros. Ultimate']
RodneyV ['Enter the Gungeon', 'Cuphead', "The Legend of Zelda: Link's Awakening", 'Celeste', 'What Remains of Edith Finch']
DaNnyDark0 ['Fire Emblem: Three Houses', 'Cuphead', 'Celeste', 'Ikaruga', 'Undertale']
AhmadAlsumain ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Super Smash Bros. Ultimate', 'Fire Emble

drumbledwarf ['Enter the Gungeon', 'Super Smash Bros. Ultimate', 'Cuphead', 'Cave Story +', 'Ikaruga']
Inzaine ['Super Mario Maker 2', 'Mario Kart 8 Deluxe', "The Legend of Zelda: Link's Awakening", 'Into the Breach', 'Celeste']
Popartdavid ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'What Remains of Edith Finch', 'Stardew Valley', 'Cuphead']
Animegroove ["The Legend of Zelda: Link's Awakening", 'Super Smash Bros. Ultimate', 'Enter the Gungeon', 'Into the Breach', 'Cave Story +']
Cygnus ['Mario Kart 8 Deluxe', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Super Smash Bros. Ultimate']
davidrp ['Super Smash Bros. Ultimate', "The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Ikaruga', 'Celeste']
EdyLee21 ['Mario Kart 8 Deluxe', 'Cave Story +', 'Ikaruga', "The Legend of Zelda: Link's Awakening", 'Enter the Gungeon']
Feco ['Super Smash Bros. Ultimate', "The Legend of Zelda: Link's Awakening", 'Cuphead', 'Enter the Gungeon', 'What Re

jimmmmo ['Cave Story +', 'Celeste', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Cuphead']
ml511341 ['Enter the Gungeon', 'Cuphead', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Into the Breach']
waddledee ['Cave Story +', 'Enter the Gungeon', 'Undertale', 'Cuphead', "The Legend of Zelda: Link's Awakening"]
EpikGamer3000 ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'What Remains of Edith Finch', 'Ikaruga', 'Into the Breach']
JasonP73 ["The Legend of Zelda: Link's Awakening", 'Cuphead', 'Enter the Gungeon', 'Cave Story +', 'Celeste']
Ave117 ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Cuphead', 'Cave Story +', 'Mario Kart 8 Deluxe']
mrpopo123 ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Donkey Kong Country: Tropical Freeze', 'Cave Story +', 'Celeste']
Lone_Wolf_Sones ['Super Mario Odyssey', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Cuphead', 'Undertale']
longoman ['Super Mario Odyssey'

Bene86 ["The Legend of Zelda: Link's Awakening", 'Cave Story +', 'Ikaruga', 'Enter the Gungeon', 'Super Smash Bros. Ultimate']
AnggeloCHL ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Super Smash Bros. Ultimate', 'Cuphead', 'Celeste']
alberttc ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Celeste', 'What Remains of Edith Finch', 'Cuphead']
SonsofZeruiah ['Celeste', 'Ikaruga', 'Stardew Valley', 'Cuphead', 'Enter the Gungeon']
Alphonsefde ['Super Smash Bros. Ultimate', 'Celeste', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Fire Emblem: Three Houses']
elPiero ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Cuphead', 'Ikaruga', 'TowerFall']
SwitchForce ["The Legend of Zelda: Link's Awakening", 'Diablo III: Eternal Collection', 'Enter the Gungeon', 'Cuphead', 'Celeste']
RockoDaFoxxo ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'What Remains of Edith Finch', 'Ikaruga', 'Celeste']
DrAdamX ['Super Smash Br

Shanesedy ['The Legend of Zelda: Breath of the Wild', "The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Ikaruga', 'What Remains of Edith Finch']
ValeryN ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Cuphead', 'Celeste', 'Undertale']
olimpo ['Enter the Gungeon', 'Undertale', 'Cuphead', 'Ikaruga', 'Fire Emblem: Three Houses']
PixelsEDM ['The Legend of Zelda: Breath of the Wild', 'Enter the Gungeon', 'Super Smash Bros. Ultimate', "The Legend of Zelda: Link's Awakening", 'Diablo III: Eternal Collection']
Juicebox-fresh ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'What Remains of Edith Finch', 'Diablo III: Eternal Collection', 'Cuphead']
Prodigic ['Super Smash Bros. Ultimate', "The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Cave Story +', 'Celeste']
brandonluhrs ['Stardew Valley', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Undertale', 'Into the Breach']
AdamSmith7 ['Enter the Gungeon', "The Legend of Zelda:

abweegeeeeeeee ['Enter the Gungeon', 'Into the Breach', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Cave Story +']
XenoX0mega ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Celeste', 'Cave Story +', 'Fire Emblem: Three Houses']
Lryu97 ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Ikaruga', 'Super Smash Bros. Ultimate', 'Cuphead']
satriaji ['Enter the Gungeon', 'The Legend of Zelda: Breath of the Wild', "The Legend of Zelda: Link's Awakening", 'What Remains of Edith Finch', 'Diablo III: Eternal Collection']
Lucky101 ['Celeste', 'Dead Cells', 'Ikaruga', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening"]
ZekeM ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Cuphead', 'Celeste', 'Cave Story +']
Catsarecool ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Cave Story +', 'Cuphead', 'Super Smash Bros. Ultimate']
pikachuandmario ['Celeste', 'Enter the Gungeon', 'Cave Story +', "The Legend of Zelda: Link's Awa

Ithenat ['Super Mario Odyssey', 'Fire Emblem: Three Houses', 'Ikaruga', 'Enter the Gungeon', 'Celeste']
Compensando ['The Legend of Zelda: Breath of the Wild', 'Diablo III: Eternal Collection', 'Enter the Gungeon', 'What Remains of Edith Finch', 'Ikaruga']
SergioJRM ['Enter the Gungeon', 'Cuphead', 'Stardew Valley', "The Legend of Zelda: Link's Awakening", 'Cave Story +']
neighborsbear ['Enter the Gungeon', 'The Legend of Zelda: Breath of the Wild', 'Mario Kart 8 Deluxe', 'Fire Emblem: Three Houses', "The Legend of Zelda: Link's Awakening"]
Franmod ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Cave Story +', 'Cuphead', 'Ikaruga']
GoPatriots ['Mario Kart 8 Deluxe', "The Legend of Zelda: Link's Awakening", 'Cuphead', 'Enter the Gungeon', 'Cave Story +']
True_Gamers ['Mario Kart 8 Deluxe', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Cuphead']
aislandener ['Enter the Gungeon', 'Undertale', 'Celeste', 'Cave Story +', "The Legend of Zelda: Link

hoppynsc ['Super Smash Bros. Ultimate', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Cave Story +', 'Dead Cells']
Hugodrummer ['Celeste', 'Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Cave Story +', 'Cuphead']
juanlobo ['The Legend of Zelda: Breath of the Wild', 'Enter the Gungeon', 'Super Mario Odyssey', "The Legend of Zelda: Link's Awakening", 'Celeste']
Wences ['Enter the Gungeon', 'Celeste', 'Dead Cells', "The Legend of Zelda: Link's Awakening", 'Cuphead']
Frikiseta ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Super Smash Bros. Ultimate', 'Cuphead']
JayMike ['Enter the Gungeon', 'Ikaruga', "The Legend of Zelda: Link's Awakening", 'Diablo III: Eternal Collection', 'Stardew Valley']
ENERGONX ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Undertale', 'Cave Story +', 'Celeste']
Collaps666 ['Enter the Gungeon', 'Cuphead', 'Super Smash Bros. Ultimate', 'Celeste', 'Ikaruga']
SwitchReview ['Downwell', 'Dead C

MJBotte1 ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Ikaruga', 'Hollow Knight']
GrungleBungle ["The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Ikaruga', 'Cave Story +', 'What Remains of Edith Finch']
Rikusorakh1 ['Super Mario Odyssey', 'Enter the Gungeon', 'Cuphead', 'FAR: Lone Sails', 'Celeste']
Asharkon ['Ikaruga', 'Enter the Gungeon', 'Super Mario Maker 2', "The Legend of Zelda: Link's Awakening", 'Stardew Valley']
Alwaysintune ['Super Mario Odyssey', 'Enter the Gungeon', 'Cave Story +', 'Celeste', 'Cuphead']
ByMineus ["The Legend of Zelda: Link's Awakening", 'Cuphead', 'Enter the Gungeon', 'Cave Story +', 'Super Smash Bros. Ultimate']
BuckeyeBrute ['Mario Kart 8 Deluxe', 'Enter the Gungeon', 'Cuphead', 'Super Smash Bros. Ultimate', 'Celeste']
GojiLink ['Super Smash Bros. Ultimate', "The Legend of Zelda: Link's Awakening", 'Undertale', 'Enter the Gungeon', 'What Remains of Edith Finch']
jzillacon ['Enter the Gungeon', 'Celeste', 'Cave St

djfuts ['Super Mario Odyssey', "The Legend of Zelda: Link's Awakening", 'Enter the Gungeon', 'Cuphead', 'Fire Emblem: Three Houses']
Zonned87 ['Enter the Gungeon', 'Ikaruga', 'Stardew Valley', 'What Remains of Edith Finch', 'Cuphead']
Simoncino ['Enter the Gungeon', 'Diablo III: Eternal Collection', "The Legend of Zelda: Link's Awakening", 'Into the Breach', 'Severed']
ZarexWolf ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Cave Story +', 'Cuphead']


In [1089]:
set(Userdf.Game_Title)

{'AI: The Somnium Files',
 'Assault Android Cactus+',
 'Astral Chain',
 'Axiom Verge',
 'Bastion',
 'Bayonetta + Bayonetta 2',
 'Bayonetta 2',
 'Cave Story +',
 'Celeste',
 'Crypt of the NecroDancer: Nintendo Switch Edition',
 'Cuphead',
 'Dead Cells',
 'Deemo',
 'Diablo III: Eternal Collection',
 'Digimon Story Cyber Sleuth: Complete Edition',
 'Divinity: Original Sin II - Definitive Edition',
 'Donkey Kong Country: Tropical Freeze',
 'Downwell',
 'Dragon Ball FighterZ',
 'Dragon Quest XI S: Echoes of an Elusive Age - Definitive Edition',
 'Dream Daddy: A Dad Dating Simulator',
 'Enter the Gungeon',
 'FAR: Lone Sails',
 'Fire Emblem: Three Houses',
 'Gone Home',
 'Guacamelee! 2',
 'Guacamelee! Super Turbo Championship Edition',
 'Hollow Knight',
 'Hyper Light Drifter',
 'INSIDE',
 'Iconoclasts',
 'Ikaruga',
 'Into the Breach',
 'Katamari Damacy REROLL',
 'Mario Kart 8 Deluxe',
 'Mark of the Ninja: Remastered',
 'Minecraft: Switch Edition',
 'Monster Boy and the Cursed Kingdom',
 'Nucl

In [1087]:
print('SamBell', [iid for (iid, _) in user_ratings])

SamBell ['Enter the Gungeon', "The Legend of Zelda: Link's Awakening", 'Celeste', 'Cave Story +', 'Cuphead']


In [1079]:
Userdf.append({'UserID': 'SamBell', 'Game_Title': 'The Legend of Zelda: Link\'s Awakening', 'Rating': 10}, ignore_index=True)

UserID                                         Game_Title  \
0            johnardo            The Legend of Zelda: Breath of the Wild   
1            triskyhd            The Legend of Zelda: Breath of the Wild   
2       silver_silver            The Legend of Zelda: Breath of the Wild   
3      EpicLadySponge            The Legend of Zelda: Breath of the Wild   
4              ocidax            The Legend of Zelda: Breath of the Wild   
5              xDeniz            The Legend of Zelda: Breath of the Wild   
6          gintoki125            The Legend of Zelda: Breath of the Wild   
7          Geraldst20            The Legend of Zelda: Breath of the Wild   
8          Sadikusama            The Legend of Zelda: Breath of the Wild   
9              jjakub            The Legend of Zelda: Breath of the Wild   
10      VideoGameDude            The Legend of Zelda: Breath of the Wild   
11          fire_fist            The Legend of Zelda: Breath of the Wild   
12           jardeson            The Legend of Zelda: Breath of the Wild   
13          grungyssj            The Legend of Zelda: Breath of the Wild   
14    AlexRamosQueiro            The Legend of Zelda: Breath of the Wild   
15       Smashmatt202            The Legend of Zelda: Breath of the Wild   
16          Prodromus            The Legend of Zelda: Breath of the Wild   
17         shawndumas            The Legend of Zelda: Breath of the Wild   
18    Fred_GameOnBatt            The Legend of Zelda: Breath of the Wild   
19     PajaritoValdes            The Legend of Zelda: Breath of the Wild   
20           Koppai15            The Legend of Zelda: Breath of the Wild   
21          Kiryu1120            The Legend of Zelda: Breath of the Wild   
22            zorakex            The Legend of Zelda: Breath of the Wild   
23          NInteNoob            The Legend of Zelda: Breath of the Wild   
24       leonardo2360            The Legend of Zelda: Breath of the Wild   
25             Xalusc            The Legend of Zelda: Breath of the Wild   
26        lesley_dias            The Legend of Zelda: Breath of the Wild   
27             altalo            The Legend of Zelda: Breath of the Wild   
28             Volian            The Legend of Zelda: Breath of the Wild   
29           LaurensL            The Legend of Zelda: Breath of the Wild   
30         flujackson            The Legend of Zelda: Breath of the Wild   
31        iusshpandeh            The Legend of Zelda: Breath of the Wild   
32      Schoolboy_hew            The Legend of Zelda: Breath of the Wild   
33           chandori            The Legend of Zelda: Breath of the Wild   
34         joseloko89            The Legend of Zelda: Breath of the Wild   
35         ChromXsome            The Legend of Zelda: Breath of the Wild   
36         The_Legend            The Legend of Zelda: Breath of the Wild   
37       beetleking22            The Legend of Zelda: Breath of the Wild   
38         Nessuno_Im            The Legend of Zelda: Breath of the Wild   
39               pg_x            The Legend of Zelda: Breath of the Wild   
40          someone98            The Legend of Zelda: Breath of the Wild   
41     FireFistYamaan            The Legend of Zelda: Breath of the Wild   
42       JuanLuisGG14            The Legend of Zelda: Breath of the Wild   
43         bimbarabim            The Legend of Zelda: Breath of the Wild   
44                  K            The Legend of Zelda: Breath of the Wild   
45             Fhranz            The Legend of Zelda: Breath of the Wild   
46            yasserb            The Legend of Zelda: Breath of the Wild   
47      darthshadow25            The Legend of Zelda: Breath of the Wild   
48    DefectiveReject            The Legend of Zelda: Breath of the Wild   
49           gaydingo            The Legend of Zelda: Breath of the Wild   
50     StefanGoedhart            The Legend of Zelda: Breath of the Wild   
51      GustavoAndres            The Legend of Zelda: Breath of

In [1084]:
Userdf.append({'UserID': 'DavidPow', 'Game_Title': 'Super Smash Bros. Ultimate', 'Rating': 10}, ignore_index=True)
Userdf.append({'UserID': 'DavidPow', 'Game_Title': 'Super Mario Odyssey', 'Rating': 9}, ignore_index=True)
Userdf.append({'UserID': 'DavidPow', 'Game_Title': 'Super Mario Party', 'Rating': 7}, ignore_index=True)

UserID                                         Game_Title  \
0            johnardo            The Legend of Zelda: Breath of the Wild   
1            triskyhd            The Legend of Zelda: Breath of the Wild   
2       silver_silver            The Legend of Zelda: Breath of the Wild   
3      EpicLadySponge            The Legend of Zelda: Breath of the Wild   
4              ocidax            The Legend of Zelda: Breath of the Wild   
5              xDeniz            The Legend of Zelda: Breath of the Wild   
6          gintoki125            The Legend of Zelda: Breath of the Wild   
7          Geraldst20            The Legend of Zelda: Breath of the Wild   
8          Sadikusama            The Legend of Zelda: Breath of the Wild   
9              jjakub            The Legend of Zelda: Breath of the Wild   
10      VideoGameDude            The Legend of Zelda: Breath of the Wild   
11          fire_fist            The Legend of Zelda: Breath of the Wild   
12           jardeson            The Legend of Zelda: Breath of the Wild   
13          grungyssj            The Legend of Zelda: Breath of the Wild   
14    AlexRamosQueiro            The Legend of Zelda: Breath of the Wild   
15       Smashmatt202            The Legend of Zelda: Breath of the Wild   
16          Prodromus            The Legend of Zelda: Breath of the Wild   
17         shawndumas            The Legend of Zelda: Breath of the Wild   
18    Fred_GameOnBatt            The Legend of Zelda: Breath of the Wild   
19     PajaritoValdes            The Legend of Zelda: Breath of the Wild   
20           Koppai15            The Legend of Zelda: Breath of the Wild   
21          Kiryu1120            The Legend of Zelda: Breath of the Wild   
22            zorakex            The Legend of Zelda: Breath of the Wild   
23          NInteNoob            The Legend of Zelda: Breath of the Wild   
24       leonardo2360            The Legend of Zelda: Breath of the Wild   
25             Xalusc            The Legend of Zelda: Breath of the Wild   
26        lesley_dias            The Legend of Zelda: Breath of the Wild   
27             altalo            The Legend of Zelda: Breath of the Wild   
28             Volian            The Legend of Zelda: Breath of the Wild   
29           LaurensL            The Legend of Zelda: Breath of the Wild   
30         flujackson            The Legend of Zelda: Breath of the Wild   
31        iusshpandeh            The Legend of Zelda: Breath of the Wild   
32      Schoolboy_hew            The Legend of Zelda: Breath of the Wild   
33           chandori            The Legend of Zelda: Breath of the Wild   
34         joseloko89            The Legend of Zelda: Breath of the Wild   
35         ChromXsome            The Legend of Zelda: Breath of the Wild   
36         The_Legend            The Legend of Zelda: Breath of the Wild   
37       beetleking22            The Legend of Zelda: Breath of the Wild   
38         Nessuno_Im            The Legend of Zelda: Breath of the Wild   
39               pg_x            The Legend of Zelda: Breath of the Wild   
40          someone98            The Legend of Zelda: Breath of the Wild   
41     FireFistYamaan            The Legend of Zelda: Breath of the Wild   
42       JuanLuisGG14            The Legend of Zelda: Breath of the Wild   
43         bimbarabim            The Legend of Zelda: Breath of the Wild   
44                  K            The Legend of Zelda: Breath of the Wild   
45             Fhranz            The Legend of Zelda: Breath of the Wild   
46            yasserb            The Legend of Zelda: Breath of the Wild   
47      darthshadow25            The Legend of Zelda: Breath of the Wild   
48    DefectiveReject            The Legend of Zelda: Breath of the Wild   
49           gaydingo            The Legend of Zelda: Breath of the Wild   
50     StefanGoedhart            The Legend of Zelda: Breath of the Wild   
51      GustavoAndres            The Legend of Zelda: Breath of